controller_tuner_nets.py

```diff
@@ -0,0 +1,117 @@
+import torch
+from torch import nn
+import numpy as np
+
+
+class ControllerV1(nn.Module):
+    def __init__(self):
+        super(ControllerV1, self).__init__()
+        self.controller = nn.Sequential(
+            nn.Linear(4 + 4 + 1, 16),
+            nn.Tanh(),
+            nn.Linear(16, 16),
+            nn.Tanh(),
+            nn.Linear(16, 16),
+            nn.Tanh(),
+            nn.Linear(16, 4),
+            nn.Softplus()
+        )
+
+    def forward(self, x):
+        x = self.controller(x)
+        return torch.exp(x)
+
+
+class ControllerExternalDerivativesV1(ControllerV1):
+    # TODO: universal Controller for all external controllers
+    def __init__(self):
+        super().__init__()
+        self.controller[0] = nn.Linear(4, 16)
+        self.controller[-1] = nn.Tanh()
+        # TODO: remove Dropout when apply -- add controller.eval() everywhere
+        # TODO: check
+
+    def forward(self, x):
+        return self.controller(x) * 50  # TODO: limit values with less crutches
+
+
+class TunerCoordinatesV1(nn.Module):
+    def __init__(self, ar=3):
+        super(TunerCoordinatesV1, self).__init__()
+        self.ar = ar
+        self.tuner = nn.Sequential(
+            nn.Linear(4 * ar, 16),
+            nn.Tanh(),
+            nn.Linear(16, 16),
+            nn.Tanh(),
+            nn.Linear(16, 16),
+            nn.Tanh(),
+            nn.Linear(16, 4),
+            nn.Tanh()
+        )
+
+    def forward(self, x):
+        # TODO: x2, y2 += 1
+        return self.tuner(x) * 2
+
+
+class TunerAnglesV1(nn.Module):
+    def __init__(self, ar=3, dropout=0.1):
+        super(TunerAnglesV1, self).__init__()
+        self.ar = ar
+        self.tuner = nn.Sequential(
+            nn.Linear(2 * ar, 16),
+            nn.Tanh(),
+            nn.Dropout(dropout),
+            nn.Linear(16, 16),
+            nn.Tanh(),
+            nn.Dropout(dropout),
+            nn.Linear(16, 16),
+            nn.Tanh(),
+            nn.Dropout(dropout),
+            nn.Linear(16, 2),
+            nn.Tanh()
+        )
+        # self._i = 0
+
+    # TODO: add length dependency
+    # TODO: пофиксить резкий переход в 0 на 360
+    # TODO: faster using torch stuff instead of numpy?
+    @staticmethod
+    def _coordinates2angles(coords):
+        coords = coords.numpy()
```

Comment:
If coords is a tensor on cuda it can cause exception. Better `coords = coords.cpu().numpy()`

Reply:
OK